File: sdk/npm/src/providers/fmail.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'fmail';
const BASE_URL = 'https://fmail.men';

interface RandomResponse {
  username?: string;
  domain?: string;
  address?: string;
}

interface InboxRow {
  id?: string | number;
  token?: string;
  sender?: string;
  subject?: string;
  snippet?: string;
  received_at?: string | number;
}

interface InboxResponse {
  emails?: InboxRow[];
}

interface EmailDetailResponse {
  token?: string;
  recipient?: string;
  domain?: string;
  sender?: string;
  subject?: string;
  body_text?: string;
  body_html?: string;
  received_at?: string | number;
}

async function fetchJSON<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      Accept: 'application/json',
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`fmail http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

function flattenMessage(raw: Record<string, unknown>, recipient: string): Record<string, unknown> {
  return {
    id: raw.id || raw.uid || raw.token || '',
    from: raw.sender || raw.from || '',
    to: raw.recipient || raw.to || recipient,
    subject: raw.subject || '',
    text: raw.body_text || raw.text || raw.snippet || '',
    html: raw.body_html || raw.html || '',
    date: raw.received_at || raw.timestamp || raw.date || '',
    attachments: raw.attachments || [],
  };
}

function normalizeDomain(domain?: string | null): string | undefined {
  const value = String(domain || '').trim();
  if (!value) {
    return undefined;
  }
  return value.replace(/^@/, '');
}

export async function generateEmail(domain?: string | null): Promise<InternalEmailInfo> {
  const selectedDomain = normalizeDomain(domain);
  const url = selectedDomain
    ? `${BASE_URL}/v1/random?domain=${encodeURIComponent(selectedDomain)}`
    : `${BASE_URL}/v1/random`;
  const json = await fetchJSON<RandomResponse>(url);
  const email = String(json?.address || '').trim();
  if (!email || !email.includes('@')) {
    throw new Error('fmail: invalid random response');
  }
  return { channel: CHANNEL, email };
}

export async function getEmails(email: string): Promise<Email[]> {
  const [local, domain] = String(email || '').split('@');
  if (!local || !domain) {
    throw new Error('fmail: invalid email');
  }

  const json = await fetchJSON<InboxResponse>(
    `${BASE_URL}/v1/inbox/${encodeURIComponent(local)}?domain=${encodeURIComponent(domain)}&limit=50`,
  );
  const rows = Array.isArray(json?.emails) ? json.emails : [];
  const out: Email[] = [];

  for (const row of rows) {
    const token = String(row?.token || row?.id || '').trim();
    if (!token) {
      out.push(normalizeEmail(flattenMessage(row as Record<string, unknown>, email), email));
      continue;
    }

    try {
      const detail = await fetchJSON<EmailDetailResponse>(`${BASE_URL}/v1/email/${encodeURIComponent(token)}`);
      out.push(normalizeEmail(flattenMessage(detail as Record<string, unknown>, email), email));
    } catch {
      out.push(normalizeEmail(flattenMessage(row as Record<string, unknown>, email), email));
    }
  }

  return out;
}

File: sdk/go/provider/fmail.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const fmailBase = "https://fmail.men"

func fmailAnyString(v any) string {
	switch x := v.(type) {
	case string:
		return strings.TrimSpace(x)
	case fmt.Stringer:
		return strings.TrimSpace(x.String())
	case nil:
		return ""
	default:
		return strings.TrimSpace(fmt.Sprint(v))
	}
}

func fmailJSON(method, u string) (map[string]any, error) {
	req, err := http.NewRequest(method, u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", GetCurrentUA())

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("fmail http %d", resp.StatusCode)
	}

	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, err
	}
	return out, nil
}

func fmailNormalizeDomain(domain *string) *string {
	if domain == nil {
		return nil
	}
	value := strings.TrimSpace(*domain)
	value = strings.TrimPrefix(value, "@")
	if value == "" {
		return nil
	}
	return &value
}

func fmailFlattenMessage(raw map[string]any, recipient string) map[string]any {
	attachments := raw["attachments"]
	if attachments == nil {
		attachments = []any{}
	}
	from := fmailAnyString(raw["sender"])
	if from == "" {
		from = fmailAnyString(raw["from"])
	}
	text := fmailAnyString(raw["body_text"])
	if text == "" {
		text = fmailAnyString(raw["text"])
	}
	if text == "" {
		text = fmailAnyString(raw["snippet"])
	}
	html := fmailAnyString(raw["body_html"])
	if html == "" {
		html = fmailAnyString(raw["html"])
	}
	date := fmailAnyString(raw["received_at"])
	if date == "" {
		date = fmailAnyString(raw["timestamp"])
	}
	if date == "" {
		date = fmailAnyString(raw["date"])
	}
	id := fmailAnyString(raw["id"])
	if id == "" {
		id = fmailAnyString(raw["uid"])
	}
	if id == "" {
		id = fmailAnyString(raw["token"])
	}
	to := fmailAnyString(raw["recipient"])
	if to == "" {
		to = fmailAnyString(raw["to"])
	}
	if to == "" {
		to = recipient
	}
	return map[string]any{
		"id":          id,
		"from":        from,
		"to":          to,
		"subject":     fmailAnyString(raw["subject"]),
		"text":        text,
		"html":        html,
		"date":        date,
		"is_read":     raw["is_read"],
		"attachments": attachments,
	}
}

func FmailGenerate(domain *string) (*CreatedMailbox, error) {
	selected := fmailNormalizeDomain(domain)
	u := fmailBase + "/v1/random"
	if selected != nil {
		u += "?domain=" + url.QueryEscape(*selected)
	}
	data, err := fmailJSON(http.MethodGet, u)
	if err != nil {
		return nil, err
	}

	email := fmailAnyString(data["address"])
	if email == "" {
		username := fmailAnyString(data["username"])
		dom := fmailAnyString(data["domain"])
		if username != "" && dom != "" {
			email = username + "@" + dom
		}
	}
	if email == "" || !strings.Contains(email, "@") {
		return nil, fmt.Errorf("fmail: invalid random response")
	}

	return &CreatedMailbox{
		Channel: "fmail",
		Email:   email,
	}, nil
}

func FmailGetEmails(email string) ([]NormEmail, error) {
	address := strings.TrimSpace(email)
	if address == "" {
		return nil, fmt.Errorf("fmail: invalid email")
	}
	parts := strings.SplitN(address, "@", 2)
	if len(parts) != 2 || strings.TrimSpace(parts[0]) == "" || strings.TrimSpace(parts[1]) == "" {
		return nil, fmt.Errorf("fmail: invalid email")
	}

	u := fmt.Sprintf("%s/v1/inbox/%s?domain=%s&limit=50", fmailBase, url.PathEscape(strings.TrimSpace(parts[0])), url.QueryEscape(strings.TrimSpace(parts[1])))
	data, err := fmailJSON(http.MethodGet, u)
	if err != nil {
		return nil, err
	}

	rows, _ := data["emails"].([]any)
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		raw, ok := row.(map[string]any)
		if !ok {
			continue
		}
		token := fmailAnyString(raw["token"])
		if token == "" {
			token = fmailAnyString(raw["id"])
		}
		if token == "" {
			out = append(out, NormalizeMap(fmailFlattenMessage(raw, address), address))
			continue
		}

		detail, err := fmailJSON(http.MethodGet, fmailBase+"/v1/email/"+url.PathEscape(token))
		if err != nil {
			out = append(out, NormalizeMap(fmailFlattenMessage(raw, address), address))
			continue
		}
		if nested, ok := detail["email"].(map[string]any); ok {
			detail = nested
		}
		out = append(out, NormalizeMap(fmailFlattenMessage(detail, address), address))
	}

	return out, nil
}
